sgkit/stats/genedrop.py

```diff
@@ -0,0 +1,344 @@
+from typing import Hashable, Optional
+
+import dask.array as da
+import numpy as np
+import xarray as xr
+from typing_extensions import Literal
+from xarray import Dataset
+
+from sgkit import variables
+from sgkit.accelerate import numba_guvectorize, numba_jit
+from sgkit.typing import ArrayLike
+from sgkit.utils import (
+    conditional_merge_datasets,
+    create_dataset,
+    define_variable_if_absent,
+)
+
+from .pedigree import (
+    _compress_hamilton_kerr_parameters,
+    parent_indices,
+    topological_argsort,
+)
+
+EST_DIPLOID = "diploid"
+EST_HAMILTON_KERR = "Hamilton-Kerr"
+
+
+@numba_guvectorize(  # type: ignore
+    [
+        "void(int8[:,:], int64[:,:], uint32, int8[:,:])",
+        "void(int16[:,:], int64[:,:], uint32, int16[:,:])",
+        "void(int32[:,:], int64[:,:], uint32, int32[:,:])",
+        "void(int64[:,:], int64[:,:], uint32, int64[:,:])",
+    ],
+    "(n,k),(n,p),()->(n,k)",
+)
+def genedrop_diploid(
+    genotypes: ArrayLike,
+    parent: ArrayLike,
+    seed: ArrayLike,
+    out: ArrayLike,
+) -> None:  # pragma: no cover
+    n_sample, n_parent = parent.shape
+    _, ploidy = genotypes.shape
+    if n_parent != 2:
+        raise ValueError("The parents dimension must be length 2")
+    if ploidy != 2:
+        raise ValueError("Genotypes are not diploid")
+    order = topological_argsort(parent)
+    np.random.seed(seed)
```

Comment:
using the global rng like this is generally a Bad Idea. 
```
rng = numpy.random.default_rng(seed)
...
idx = rng.randint(2)
```
is a much better pattern. 

Does numba support doing it like this? If not, might be simplest to pass in a per-generated array of random values (even if you don't use them all)

Reply:
Sorry for slow the response @jeromekelleher. Numba doesn't seem to support `numpy.random.default_rng`. For reference, the [docs](https://numba.pydata.org/numba-doc/dev/reference/numpysupported.html#random) state "Since version 0.28.0, the generator is thread-safe and fork-safe. Each thread and each process will produce independent streams of random numbers."

Reply:
That's very tricky then - it's super important that simulators are deterministic, so you can reproduce weird problems later. If there's a reasonable upper limit on the number of random values you need, I'd strongly recommend generating those beforehand with numpy and consuming within the jitted function.

Reply:
As it stands, the simulations are deterministic if a random seed is specified by the user. If not, then a seed is randomly chosen and used to generate random seeds for the simulation to ensure that the resulting dask array produces the same result when re-evaluated. Would you preference be to require a random seed is given by the user?